daml-lf/spec/daml-lf-1.rst

```diff
@@ -3151,68 +3134,144 @@ in which keys were inserted into the map (insertion-order).
+    —————————————————————————————————————————————————————————————————————— EvGenMapInsertReplaceErr
+      𝕆('GENMAP_INSERT' @σ @τ 〚v₁ ↦ w₁; … ; vₙ ↦ wₙ〛 v w) = Err t
+
+      𝕆('EQUAL' @σ vᵢ v) = Ok 'True'    for some i ∈ 1, …, n
+    —————————————————————————————————————————————————————————————————————— EvGenMapInsertReplace
+      𝕆('GENMAP_INSERT' @σ @τ 〚v₁ ↦ w₁; …; vₙ ↦ wₙ〛 v w) =
+        'Ok' 〚v₁ ↦ w₁; …; vᵢ₋₁ ↦ wᵢ₋₁; vᵢ ↦ w;  vᵢ₊₁ ↦ wᵢ₊₁; …; vₙ ↦ wₙ〛
 
-  This raises an error if the key is not a valid map key. Keys are
-  compared according to the rules listed below.
+      𝕆('LESS' @σ vᵢ₋₁ v) = Err t    for some i ∈ 1, …, n
+    —————————————————————————————————————————————————————————————————————— EvGenMapInsertInsertErrLeft
+      𝕆('GENMAP_INSERT' @σ @τ 〚v₁ ↦ w₁; …; vₙ ↦ wₙ〛 v w) = Err t
+
+      𝕆('LESS' @σ v vᵢ) = Err t    for some i ∈ 1, …, n
+    —————————————————————————————————————————————————————————————————————— EvGenMapInsertInsertErrRight
+      𝕆('GENMAP_INSERT' @σ @τ 〚v₁ ↦ w₁; …; vₙ ↦ wₙ〛 v w) = Err t
+
+      𝕆('LESS' @σ v v₁) = Ok 'True'
+    —————————————————————————————————————————————————————————————————————— EvGenMapInsertInsertFirst
+      𝕆('GENMAP_INSERT' @σ @τ 〚v₁ ↦ w₁; …; vₙ ↦ wₙ〛 v w) =
+        'Ok' 〚v ↦ w; v₁ ↦ w₁; …; vₙ ↦ wₙ〛
+
+      𝕆('LESS' @σ vᵢ₋₁ v) = Ok 'True'
+      𝕆('LESS' @σ v vᵢ) = Ok 'True'
+      for some i ∈ 2, …, n-1
+    —————————————————————————————————————————————————————————————————————— EvGenMapInsertInsertMiddle
+      𝕆('GENMAP_INSERT' @σ @τ 〚v₁ ↦ w₁; … ; vₙ ↦ wₙ〛 v w) =
+        'Ok' 〚v₁ ↦ w₁; … ; vᵢ₋₁ ↦ wᵢ₋₁; v ↦ w;  vᵢ ↦ wᵢ; … ; vₙ ↦ wₙ〛
+
+      𝕆('LESS' @σ vₙ v) = Ok 'True'
+    —————————————————————————————————————————————————————————————————————— EvGenMapInsertInsertLast
+      𝕆('GENMAP_INSERT' @σ @τ 〚v₁ ↦ w₁; …; vₙ ↦ wₙ〛 v w) =
+        'Ok' 〚v₁ ↦ w₁; …; vₙ ↦ wₙ; v ↦ w〛
+
+
+* ``GENMAP_LOOKUP : ∀ α. ∀ β.  α → 'GenMap' α β → 'Optional' α``
+
+  Looks up the value at a key in the map using the builtin function
+  ``EQUAL`` to test key equality. This raises a runtime error if it
+  try to compare incomparable values.
 
   [*Available in versions >= 1.dev*]
 
-* ``GENMAP_DELETE : ∀ α. ∀ β.  α → 'GenMap' α β → 'GenMap' α β``
+  Formally the builtin function ``GENMAP_LOOKUP`` semantics is defined
+  by the following rules. ::
 
-  Deletes a key and its value from the map. When the key is not a
-  member of the map, the original map is returned.
+      𝕆('EQUAL' @σ vᵢ v) = Err t  for some i ∈ 1, …, n
+    —————————————————————————————————————————————————————————————————————— EvGenMapLookupErr
+      𝕆('GENMAP_LOOKUP' @σ @τ 〚v₁ ↦ w₁; … ; vₙ ↦ wₙ〛 v) = Err t
+
+      𝕆('EQUAL' @σ vᵢ v) = Ok 'True'  for some i ∈ 1, …, n
+    —————————————————————————————————————————————————————————————————————— EvGenMapLookupPresent
+      𝕆('GENMAP_LOOKUP' @σ @τ 〚v₁ ↦ w₁; … ; vₙ ↦ wₙ〛 v) =
+        'Ok' (Some wᵢ)
+
+      𝕆('EQUAL' @σ vᵢ v) = Ok 'False'  for any i ∈ 1, …, n
```

Comment:
```suggestion
      𝕆('EQUAL' @σ vᵢ v) = Ok 'False'  for all i ∈ 1, …, n
```

But I think we should also add the requirement that `n >= 1`, and add a separate rules for lookups in empty maps to ensure that a lookup with bad key is caught.